Cabal/src/Distribution/Simple/GHC/Build/ExtraSources.hs

```diff
@@ -86,7 +102,21 @@ buildCSources mbMainFile =
 buildCxxSources mbMainFile =
   buildExtraSources
     "C++ Sources"
-    Internal.componentCxxGhcOptions
+    ( \verbosity lbi bi clbi odir filename ->
+        (Internal.sourcesGhcOptions verbosity lbi bi clbi odir filename)
+          { ghcOptCcOptions =
+              ( case compilerCompatVersion GHC (compiler lbi) of
+                  Just v
+                    | v >= mkVersion [8, 10] -> Internal.defaultGhcOptCcOptions lbi bi
```

Comment:
why this is not in `sourcesGhcOptions `. It has access to `lbi`?

Reply:
It does not completely match, there are different `defaultGhcOptCcOptions` and `defaultGhcOptCxxOptions`.
You are right, `ghcOptCcOptions` and `ghcOptCxxOptions` can be moved to `sourcesGhcOptions`.
On the other hand, `ghcOptCcProgram` should be left here, since not everywhere (it seems to me) there may be access to `gccProgram`.

Reply:
> here may be access to gccProgram.

That's a bug on its own:
- If `--with-gcc` is given, it should always be passed to GHC (so GHC always consistently uses the same GCC).

Reply:
FWIW, here in `buildCxxSources ` you use `defaultGhcOptCcOptions `, and in `buildCSources` you set `defaultGhcOptCxxOptions `.

That's just proves my point, subtle differences potentially leading to bugs.

Reply:
> I only see a downsides of using subtly different calls to GHC when compiling C or C++ files.

I don't understand, in 8.10 there was a flag split, cabal decided to support versions older than 8.10 too. Maybe you think it's a bug?

> That's just proves my point, subtle differences potentially leading to bugs.

This is because we maintain the same behavior for versions older than 8.10.

Reply:
If you think this is a bug, then we need to remove this test.

`PackageTests/FFI/ForeignOptsCxx`

Reply:
And arguably we should check for supported GHC version in one place.

I'm not sure where it's done now, at GhcOptions creation time or its serialisation time.

Either way, if some options are dropped because of GHC not supporting them, they shouldn't be dropped silently.

Reply:
No. Am I explaining myself unclearly. Let's try again.

it looks like that

- You set C options and `ghcOptCcProgram` **only** in `buildCSources`
- You set C++ options and `ghcOptCxxProgram` **only** in `buildCxxSources`

That cannot be right. You should set these options **always**.

Reply:
In the retrospective, `cabal-install` could refused to support an accidental "but it did work before with older GHC, cc-options were passed to C compiler" when proper `cxx-options` were added. Those accidents just make code base very spagetti.

Reply:
It seems I have implemented everything that is written here? With reservations for versions below 9.4.